File: CMPS3500_MiniScheme_Starter_Package/procedural/tokenizer.cpp
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga  /
  / ASGT: CHECKPOINT 1                                  /
  / ORGN: CSUB - CMPS 3500                              /
  / FILE: tokenizer.cpp                                 /
  / DATE: 03/29/2026                                    /
*/

#include "tokenizer.h"
#include <cctype>

std::vector<std::string> tokenize(const std::string& source_code) {
    std::vector<std::string> token_list;
    std::string current = "";

    for (char c : source_code) {

        // Option 1: the current character is a PARENTHESIS
        if (c == '(' || c == ')') {

            // Save current token if it exists
            if (!current.empty()) {
                token_list.push_back(current);
                current.clear();
            }

            // Add the parenthesis as its own token
            token_list.push_back(std::string(1, c));
        }

        // Option 2: the current character is just WHITESPACE
        else if (std::isspace(c)) {

            // if you have whitespace, then everything before it
            // should be a single token stored in "current"
            if (!current.empty()) {
                token_list.push_back(current);
                current.clear();
            }
        }

        // Option 3: you have a token that could be one or more
        // characters long. Keep storing those characters in a single
        // token
        else {
            current += c;
        }
    }

    // Push last token if needed
    if (!current.empty()) {
        token_list.push_back(current);
    }

    return token_list;
}

File: CMPS3500_MiniScheme_Starter_Package/oop/Tokenizer.java
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga /
  / ASGT: CHECKPOINT 2                                  /
  / ORGN: CSUB - CMPS 3500                              /
  / FILE: Tokenizer.java                                /
  / DATE: 04/11/2026                                    /
*/
import java.util.ArrayList;
import java.util.List;

public class Tokenizer {
    public static List<String> tokenize(String sourceCode){
        List<String> tokenList = new ArrayList<String>();
        String current = "";

        //java cant iterate through strings like c++ can must convert to char array
        for(char c : sourceCode.toCharArray()){
            if(c == 'c' || c == ')'){
                if (!current.isEmpty()){
                    tokenList.add(current);
                    current = "";
                }
                
                tokenList.add(String.valueOf(c));
            }
            else if(Character.isWhitespace(c)){
                if(current.isEmpty()){
                    tokenList.add(current);
                    current = "";
                }
            }
            else{
                current += c;
            }
        }

        if(!current.isEmpty()){
            tokenList.add(current);
        }

        return tokenList;
    }
}